### PR TITLE
ARROW-3649: [Rust] Refactor MutableBuffer's resize

### PR DESCRIPTION
**Pull request by sunchao**

### CODE REVIEW

File: rust/src/buffer.rs [hunk 1]
```diff
@@ -196,23 +187,43 @@ impl MutableBuffer {
...
-    pub fn resize(&mut self, new_capacity: usize) -> Result<usize> {
-        if new_capacity <= self.capacity {
-            return Ok(0);
+    /// If `new_len` is greater than `len`, the buffer's length is simply adjusted to be
+    /// the former, optionally extending the capacity. The data between `len` and
+    /// `new_len` will remain unchanged.
+    ///
+    /// If `new_len` is less than `len`, the buffer will be truncated.
+    pub fn resize(&mut self, new_len: usize) -> Result<()> {
+        if new_len > self.len {
```
Comment (paddyhoran) [inline rust/src/buffer.rs]: I think this check is redundant as the call to `reserve` will check `new_len` against `self.capacity`

File: rust/src/builder.rs [hunk 2]
```diff
@@ -61,37 +61,28 @@ macro_rules! impl_buffer_builder {
...
-            /// elements of type T
-            fn grow(&mut self, new_capacity: i64) -> Result<()> {
                 let byte_capacity = mem::size_of::<$native_ty>() * new_capacity as usize;
-                self.buffer.resize(byte_capacity)?;
+                self.buffer.reserve(byte_capacity)?;
                 Ok(())
             }
 
-            /// Build an immutable `Buffer` from the existing internal `MutableBuffer`'s memory
+            /// Consumes this and returns an immutable `Buffer`.
```
Comment (paddyhoran) [inline rust/src/builder.rs]: Is there a missing word here?  i.e. Consumes this `BufferBuilder` and...

**Review by paddyhoran — COMMENTED** [review]
Looks good, couple of nits, thanks @sunchao

File: rust/src/buffer.rs (hunk 1 above)
Comment (paddyhoran) [reply rust/src/buffer.rs]: I marked the wrong line, I mean this check:
`if new_len > self.capacity {`
on line 214

Comment (sunchao) [reply rust/src/buffer.rs]: Good point. Removed.

File: rust/src/builder.rs (hunk 2 above)
Comment (sunchao) [reply rust/src/builder.rs]: Done.

**Review by kszucs — APPROVED** [review]
LGTM, Thanks @sunchao @paddyhoran

**Pull request closed without merging**